README.md
# Words To Numbers

Dart implementation of the [words_to_numbers](https://github.com/finnfiddle/words-to-numbers) package by [@finnfiddle](https://github.com/finnfiddle)

Converts words to numbers. Optionally fuzzy match the words to numbers.

The returned value is always return a string. If you are expecting the entire string to be a number, then you can use the `tryParse` methods to get the numeric value:

```dart
import 'package:words_to_numbers/words_to_numbers.dart';

void main() {
    final String numberAsText = wordsToNumbers('three point five two');
    final int intValue = int.tryParse(numberAsText); // Returns null
    final double doubleValue = double.tryParse(numberAsText); // Returns 3.52
    final num numValue = num.tryParse(numberAsText); // Returns 3.45
}
```

## Basic Examples

```dart
import 'package:words_to_numbers/words_to_numbers.dart';
void main() {
    wordsToNumbers('one hundred'); //100
    wordsToNumbers('one hundred and five'); //105
    wordsToNumbers('one hundred and twenty five'); //125
    wordsToNumbers('four thousand and thirty'); //4030
    wordsToNumbers('six million five thousand and two'); //6005002
    wordsToNumbers('a thousand one hundred and eleven'); //1111
    wordsToNumbers('twenty thousand five hundred and sixty nine'); //20569
    wordsToNumbers('five quintillion'); //5000000000000000000
    wordsToNumbers('one-hundred'); //100
    wordsToNumbers('one-hundred and five'); //105
    wordsToNumbers('one-hundred and twenty-five'); //125
    wordsToNumbers('four-thousand and thirty'); //4030
    wordsToNumbers('six-million five-thousand and two'); //6005002
    wordsToNumbers('a thousand, one-hundred and eleven'); //1111
    wordsToNumbers('twenty-thousand, five-hundred and sixty-nine'); //20569
}
```

## Multiple numbers in a string

Returns a string with all instances replaced.

```dart
wordsToNumbers('there were twenty-thousand, five-hundred and sixty-nine X in the five quintillion Y')) // 'there were 20569 X in the 5000000000000000000 Y'
```

## With Fuzzy Matching

Uses [Jaro distance](https://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance) provided by the [edit_distance](https://pub.dev/packages/edit_distance) package to find the best match for the number words. Don't rely on this being completely accurate...

```dart
import 'package:words_to_numbers/words_to_numbers.dart';
void main() {
    wordsToNumbers('won huntred', Options(fuzzy: true)); //100
    wordsToNumbers('too thousant and fiev', Options(fuzzy: true)); //2005
    wordsToNumbers('tree millyon sefen hunderd and twinty sex', Options(fuzzy: true)); //3000726
}
```

## Decimal Points

```dart
import 'package:words_to_numbers/words_to_numbers.dart';
void main() {
    wordsToNumbers('ten point five'); //10.5
    wordsToNumbers('three point one four one five nine two six'); //3.1415926
}
```

## Ordinal Numbers

```dart
import 'package:words_to_numbers/words_to_numbers.dart';
void main() {
    wordsToNumbers('first'); //1
    wordsToNumbers('second'); //2
    wordsToNumbers('third'); //3
    wordsToNumbers('fourteenth'); //14
    wordsToNumbers('twenty fifth'); //25
    wordsToNumbers('thirty fourth'); //34
    wordsToNumbers('forty seventh'); //47
    wordsToNumbers('fifty third'); //53
    wordsToNumbers('sixtieth'); //60
    wordsToNumbers('seventy second'); //72
    wordsToNumbers('eighty ninth'); //89
    wordsToNumbers('ninety sixth'); //96
    wordsToNumbers('one hundred and eighth'); //108
    wordsToNumbers('one hundred and tenth'); //110
    wordsToNumbers('one hundred and ninety ninth'); //199
}
```

## Commonjs

```dart
import 'package:words_to_numbers/words_to_numbers.dart';
wordsToNumbers('one hundred'); //100;
```

## Implied Hundreds

```dart
import 'package:words_to_numbers/words_to_numbers.dart';
void main() {
    wordsToNumbers('nineteen eighty four', {Options(impliedHundreds: true)); //1984
    wordsToNumbers('one thirty', {Options(impliedHundreds: true)); //130
    wordsToNumbers('six sixty two', {Options(impliedHundreds: true)); //662
    wordsToNumbers('ten twelve', {Options(impliedHundreds: true)); //1012
    wordsToNumbers('nineteen ten', {Options(impliedHundreds: true)); //1910
    wordsToNumbers('twenty ten', {Options(impliedHundreds: true)); //2010
    wordsToNumbers('twenty seventeen', {Options(impliedHundreds: true)); //2017
    wordsToNumbers('twenty twenty', {Options(impliedHundreds: true)); //2020
    wordsToNumbers('twenty twenty one', {Options(impliedHundreds: true)); //2021
    wordsToNumbers('fifty sixty three', {Options(impliedHundreds: true)); //5063
    wordsToNumbers('fifty sixty', {Options(impliedHundreds: true)); //5060
    wordsToNumbers('fifty sixty three thousand', {Options(impliedHundreds: true)); //5063000
    wordsToNumbers('one hundred thousand', {Options(impliedHundreds: true)); //100000
}
```

CHANGELOG.md
# Changelog

## 1.0.0 - 22/07/20

* Initial release 


test/words_to_numbers_test.dart
import 'package:words_to_numbers/words_to_numbers.dart';
import 'package:test/test.dart';

void main() {

  final wtn = wordsToNumbers;

  test('one hundred', () {
    expect(wtn('one hundred'), '100');
  });

  test('one hundred two', () {
    expect(wtn('one hundred two'), '102');
  });

  test('one hundred and five', () {
    expect(wtn('one hundred and five'), '105');
  });

  test('one hundred and twenty five', () {
    expect(wtn('one hundred and twenty five'), '125');
  });

  test('four thousand and thirty', () {
    expect(wtn('four thousand and thirty'), '4030');
  });

  test('six million five thousand and two', () {
    expect(wtn('six million five thousand and two'), '6005002');
  });

  test('a thousand one hundred and eleven', () {
    expect(wtn('a thousand one hundred and eleven'), '1111');
  });

  test('sixty nine', () {
    expect(wtn('sixty nine'), '69');
  });

  test('twenty thousand five hundred and sixty nine', () {
    expect(wtn('twenty thousand five hundred and sixty nine'), '20569');
  });

  test('five quintillion', () {
    expect(wtn('five quintillion'), '5000000000000000000');
  });

  test('one-hundred', () {
    expect(wtn('one-hundred'), '100');
  });

  test('one-hundred and five', () {
    expect(wtn('one-hundred and five'), '105');
  });

  test('one-hundred and twenty-five', () {
    expect(wtn('one-hundred and twenty-five'), '125');
  });

  test('four-thousand and thirty', () {
    expect(wtn('four-thousand and thirty'), '4030');
  });

  test('six-million five-thousand and two', () {
    expect(wtn('six-million five-thousand and two'), '6005002');
  });

  test('a thousand, one-hundred and eleven', () {
    expect(wtn('a thousand, one-hundred and eleven'), '1111');
  });

  test('twenty-thousand, five-hundred and sixty-nine', () {
    expect(wtn('twenty-thousand, five-hundred and sixty-nine'), '20569');
  });

  test('there were twenty-thousand, five-hundred and sixty-nine X in the five quintillion Y', () {
    expect(wtn('there were twenty-thousand, five-hundred and sixty-nine X in the five quintillion Y'), 'there were 20569 X in the 5000000000000000000 Y');
  });

  test('one two three', () {
    expect(wtn('one two three'), '1 2 3');
  });

  test('test one two three test', () {
    expect(wtn('test one two three test'), 'test 1 2 3 test');
  });

  test('tu thousant and faav', () {
    expect(wtn('too thousant and fiev', Options(fuzzy: true)), '2005');
  });

  // The following tests has been changed slightly for Dart, as the Jaro-Winkler
  // distance implementation used performs differently on strings with 3 or fewer letters
  test('onf huntred', () {
    expect(wtn('onf huntred', Options(fuzzy: true)), '100');
  });

  test('tree millyon sefen hunderd ant twinty sex', () {
    expect(wtn('tree millyon sefen hunderd and twinty sex', Options(fuzzy: true)), '3000726');
  });

  test('forty two point five', () {
    expect(wtn('forty two point five'), '42.5');
  });

  test('ten point five', () {
    expect(wtn('ten point five'), '10.5');
  });

  test('three point one four one five nine two six', () {
    expect(wtn('three point one four one five nine two six'), '3.1415926');
  });

  /* testing for ordinal numbers */

  test('first', () {
    expect(wtn('first'), '1');
  });

  test('second', () {
    expect(wtn('second'), '2');
  });

  test('third', () {
    expect(wtn('third'), '3');
  });

  test('fourteenth', () {
    expect(wtn('fourteenth'), '14');
  });

  test('twenty fifth', () {
    expect(wtn('twenty fifth'), '25');
  });

  test('thirty fourth', () {
    expect(wtn('thirty fourth'), '34');
  });

  test('forty seventh', () {
    expect(wtn('forty seventh'), '47');
  });

  test('fifty third', () {
    expect(wtn('fifty third'), '53');
  });

  test('sixtieth', () {
    expect(wtn('sixtieth'), '60');
  });

  test('seventy second', () {
    expect(wtn('seventy second'), '72');
  });

  test('eighty ninth', () {
    expect(wtn('eighty ninth'), '89');
  });

  test('ninety sixth', () {
    expect(wtn('ninety sixth'), '96');
  });

  test('one hundred and eighth', () {
    expect(wtn('one hundred and eighth'), '108');
  });

  test('one hundred and tenth', () {
    expect(wtn('one hundred and tenth'), '110');
  });

  test('one hundred and ninety ninth', () {
    expect(wtn('one hundred and ninety ninth'), '199');
  });

  test('digtest one', () {
    expect(wtn('digtest one'), 'digtest 1');
  });

  test('digtest one ', () {
    expect(wtn('digtest one '), 'digtest 1 ');
  });

  test('one thirty', () {
    expect(wtn('one thirty'), '1 30');
  });

  test('thousand', () {
    expect(wtn('thousand'), '1000');
  });

  test('million', () {
    expect(wtn('million'), '1000000');
  });

  test('billion', () {
    expect(wtn('billion'), '1000000000');
  });

  test('xxxxxxx one hundred', () {
    expect(wtn('xxxxxxx one hundred'), 'xxxxxxx 100');
  });

  test('and', () {
    expect(wtn('and'), 'and');
  });

  test('a', () {
    expect(wtn('a'), 'a');
  });

  test('junkvalue', () {
    expect(wtn('junkvalue'), 'junkvalue');
  });

  test('eleven dot one', () {
    expect(wtn('eleven dot one'), '11.1');
  });

  test('Fifty People, One Question: Brooklyn', () {
    expect(wtn('Fifty People, One Question: Brooklyn'), '50 People, 1 Question: Brooklyn');
  });

  test('Model Fifty-One Fifty-Six', () {
    expect(wtn('Model Fifty-One Fifty-Six'), 'Model 51 56');
  });

  test('Fifty Million Frenchmen', () {
    expect(wtn('Fifty Million Frenchmen'), '50000000 Frenchmen');
  });

  test('A Thousand and One Wives', () {
    expect(wtn('A Thousand and One Wives'), '1001 Wives');
  });

  test('Ten Thousand Pictures of You', () {
    expect(wtn('Ten Thousand Pictures of You'), '10000 Pictures of You');
  });

  test('nineteen eighty four', () {
    expect(wtn('nineteen eighty four', Options(impliedHundreds: true)), '1984');
  });

  test('one thirty', () {
    expect(wtn('one thirty', Options(impliedHundreds: true)), '130');
  });

  test('six sixty two', () {
    expect(wtn('six sixty two', Options(impliedHundreds: true)), '662');
  });

  test('ten twelve', () {
    expect(wtn('ten twelve', Options(impliedHundreds: true)), '1012');
  });

  test('nineteen ten', () {
    expect(wtn('nineteen ten', Options(impliedHundreds: true)), '1910');
  });

  test('twenty ten', () {
    expect(wtn('twenty ten', Options(impliedHundreds: true)), '2010');
  });

  test('twenty seventeen', () {
    expect(wtn('twenty seventeen', Options(impliedHundreds: true)), '2017');
  });

  test('twenty twenty', () {
    expect(wtn('twenty twenty', Options(impliedHundreds: true)), '2020');
  });

  test('twenty twenty one', () {
    expect(wtn('twenty twenty one', Options(impliedHundreds: true)), '2021');
  });

  test('fifty sixty three', () {
    expect(wtn('fifty sixty three', Options(impliedHundreds: true)), '5063');
  });

  test('fifty sixty', () {
    expect(wtn('fifty sixty', Options(impliedHundreds: true)), '5060');
  });

  test('three thousand', () {
    expect(wtn('three thousand', Options(impliedHundreds: true)), '3000');
  });

  test('fifty sixty three thousand', () {
    expect(wtn('fifty sixty three thousand', Options(impliedHundreds: true)), '5063000');
  });

  test('one hundred thousand', () {
    expect(wtn('one hundred thousand'), '100000');
  });

  test('I have zero apples and four oranges', () {
    expect(wtn('I have zero apples and four oranges'), 'I have 0 apples and 4 oranges');
  });

  test('Dot two Dot', () {
    expect(wtn('Dot two Dot'), '0.2 Dot');
  });

  test('seventeen dot two four dot twelve dot five', () {
    expect(wtn('seventeen dot two four dot twelve dot five'), '17.24 dot 12.5');
  });

  // Added tests for decimals with non-unit expression
  test('point hundred and three', () {
    expect(wtn('point hundred and three'), '0.103');
  });

  test('point seventy-nine', () {
    expect(wtn('point seventy-nine'), '0.79');
  });

// Known failing cases

  // test('one thirty thousand', () {
  //   expect(wtn('one thirty thousand', Options(impliedHundreds: true)), '130000');
  // });

  // test('nineteen eighty thousand', () {
  //   expect(wtn('nineteen eighty thousand', Options(impliedHundreds: true )), '19 80000');
  // });

  // test('one hundred two thousand', () {
  //   expect(wtn('one hundred two thousand'), '102000');
  // });

  // test('one hundred and two thousand', () {
  //   expect(wtn('one hundred and two thousand'), '102000');
  // });

}



lib/src/parser.dart
import 'package:meta/meta.dart';

import 'constants.dart';
import 'fuzzy.dart';
import 'options.dart';


enum TokenFit {
  SKIP, 
  ADD, 
  START_NEW_REGION, 
  NOPE
}

bool canAddTokenToEndOfSubRegion(SubRegion subRegion, Chunk currentToken, bool impliedHundreds) {
  final tokens = subRegion.tokens;
  final prevToken = tokens[0];
  if (prevToken == null) return true;
  if (
    prevToken.type == TOKEN_TYPE.MAGNITUDE &&
    currentToken.type == TOKEN_TYPE.UNIT
  ) return true;
  if (
    prevToken.type == TOKEN_TYPE.MAGNITUDE &&
    currentToken.type == TOKEN_TYPE.TEN
  ) return true;
  if (
    impliedHundreds &&
    subRegion.type == TOKEN_TYPE.MAGNITUDE &&
    prevToken.type == TOKEN_TYPE.TEN &&
    currentToken.type == TOKEN_TYPE.UNIT
  ) return true;
  if (
    impliedHundreds &&
    subRegion.type == TOKEN_TYPE.MAGNITUDE &&
    prevToken.type == TOKEN_TYPE.UNIT &&
    currentToken.type == TOKEN_TYPE.TEN
  ) return true;
  if (
    prevToken.type == TOKEN_TYPE.TEN &&
    currentToken.type == TOKEN_TYPE.UNIT
  ) return true;
  if (
    !impliedHundreds &&
    prevToken.type == TOKEN_TYPE.TEN &&
    currentToken.type == TOKEN_TYPE.UNIT
  ) return true;
  if (
    prevToken.type == TOKEN_TYPE.MAGNITUDE &&
    currentToken.type == TOKEN_TYPE.MAGNITUDE
  ) return true;
  if (
    !impliedHundreds &&
    prevToken.type == TOKEN_TYPE.TEN &&
    currentToken.type == TOKEN_TYPE.TEN
  ) return false;
  if (
    impliedHundreds &&
    prevToken.type == TOKEN_TYPE.TEN &&
    currentToken.type == TOKEN_TYPE.TEN
  ) return true;
  return false;
}

SubRegionType getSubRegionType(SubRegion subRegion, Chunk currentToken) {
  if (subRegion == null) {
    return SubRegionType(type: currentToken.type);
  }
  final prevToken = subRegion.tokens[0];
  final isHundred = (
    (prevToken.type == TOKEN_TYPE.TEN && currentToken.type == TOKEN_TYPE.UNIT) ||
    (prevToken.type == TOKEN_TYPE.TEN && currentToken.type == TOKEN_TYPE.TEN) ||
    (
      prevToken.type == TOKEN_TYPE.UNIT && currentToken.type == TOKEN_TYPE.TEN &&
      NUMBER[prevToken.lowerCaseValue] > 9
    ) ||
    (prevToken.type == TOKEN_TYPE.UNIT && currentToken.type == TOKEN_TYPE.UNIT) ||
    (prevToken.type == TOKEN_TYPE.TEN && currentToken.type == TOKEN_TYPE.UNIT && subRegion.type == TOKEN_TYPE.MAGNITUDE)
  );
  if (currentToken.type == TOKEN_TYPE.DECIMAL) return SubRegionType(type: TOKEN_TYPE.DECIMAL, isHundred: isHundred);
  if (subRegion.type == TOKEN_TYPE.MAGNITUDE) return SubRegionType(type: TOKEN_TYPE.MAGNITUDE, isHundred: isHundred);
  if (isHundred) return SubRegionType( type: TOKEN_TYPE.HUNDRED, isHundred: isHundred);
  return SubRegionType( type: currentToken.type, isHundred: isHundred);
}

class SubRegionType {
  final TOKEN_TYPE type;
  final bool isHundred;

  SubRegionType({this.type, this.isHundred = false});
}

SubRegionFit checkIfTokenFitsSubRegion(SubRegion subRegion, Chunk token, Options options) {
  final subType = getSubRegionType(subRegion, token);
  final type = subType.type;
  final isHundred = subType.isHundred;
  if (subRegion == null) return SubRegionFit(action: TokenFit.START_NEW_REGION, type: type, isHundred: isHundred);
  if (canAddTokenToEndOfSubRegion(subRegion, token, options?.impliedHundreds)) {
    return SubRegionFit(action: TokenFit.ADD, type: type, isHundred: isHundred);
  }
  return SubRegionFit(action: TokenFit.START_NEW_REGION, type: type, isHundred: isHundred);
}

class SubRegionFit {
  final TokenFit action;
  final TOKEN_TYPE type;
  final bool isHundred;

  SubRegionFit({this.action, this.type, this.isHundred = false});
}

List<SubRegion> getSubRegions(Region region, Options options) {
  final subRegions = List<SubRegion>();
  SubRegion currentSubRegion;
  final tokensCount = region.tokens.length;
  int i = tokensCount - 1;
  while (i >= 0) {
    final token = region.tokens[i];
    final subRegionFit = checkIfTokenFitsSubRegion(currentSubRegion, token, options);
    final action = subRegionFit.action, type = subRegionFit.type, isHundred = subRegionFit.isHundred;
    token.type = isHundred ? TOKEN_TYPE.HUNDRED : token.type;
    switch (action) {
      case TokenFit.ADD: {
        currentSubRegion.type = type;
        currentSubRegion.tokens.insert(0, token);
        break;
      }
      case TokenFit.START_NEW_REGION: {
        currentSubRegion = SubRegion(
          tokens: [token],
          type: type,
        );
        subRegions.insert(0, currentSubRegion);
        break;
      }
      default: 
      // no action
    }
    i--;
  }
  return subRegions;
}

bool canAddTokenToEndOfRegion(Region region, Chunk currentToken, bool impliedHundreds) {
  final tokens = region.tokens;
  final prevToken = tokens[tokens.length - 1];
  if (
    !impliedHundreds &&
    prevToken.type == TOKEN_TYPE.UNIT &&
    currentToken.type == TOKEN_TYPE.UNIT &&
    !region.hasDecimal
  ) return false;
  if (!impliedHundreds && prevToken.type == TOKEN_TYPE.UNIT && currentToken.type == TOKEN_TYPE.TEN) return false;
  if (!impliedHundreds && prevToken.type == TOKEN_TYPE.TEN && currentToken.type == TOKEN_TYPE.TEN) return false;
  return true;
}

TokenFit checkIfTokenFitsRegion(Region region, Chunk token, Options options) {
  final isDecimal = DECIMALS.contains(token.lowerCaseValue);
  if ((region == null || region.tokens.isEmpty) && isDecimal) {
    return TokenFit.START_NEW_REGION;
  }
  final isPunctuation = PUNCTUATION.contains(token.lowerCaseValue);
  if (isPunctuation) return TokenFit.SKIP;
  final isJoiner = JOINERS.contains(token.lowerCaseValue);
  if (isJoiner) return TokenFit.SKIP;
  if (isDecimal && !region.hasDecimal) {
    return TokenFit.ADD;
  }
  final isNumberWord = NUMBER_WORDS.contains(token.lowerCaseValue);
  if (isNumberWord) {
    if (region == null) return TokenFit.START_NEW_REGION;
    if (canAddTokenToEndOfRegion(region, token, options?.impliedHundreds)) {
      return TokenFit.ADD;
    }
    return TokenFit.START_NEW_REGION;
  }
  return TokenFit.NOPE;
}

bool checkBlacklist(List<Chunk> tokens) {
  return tokens.length == 1 && BLACKLIST_SINGULAR_WORDS.contains(tokens[0].lowerCaseValue);
}
  
List<Region> matchRegions(List<Chunk> tokens, options) {
  final regions = List<Region>();

  if (checkBlacklist(tokens)) return regions;

  int i = 0;
  Region currentRegion;
  final tokensCount = tokens.length;
  while (i < tokensCount) {
    final token = tokens[i];
    final tokenFits = checkIfTokenFitsRegion(currentRegion, token, options);
    switch (tokenFits) {
      case TokenFit.SKIP: {
        break;
      }
      case TokenFit.ADD: {
        if (currentRegion != null) {
          currentRegion.end = token.end;
          currentRegion.tokens.add(token);
          if (token.type == TOKEN_TYPE.DECIMAL) {
            currentRegion.hasDecimal = true;
          }
        }
        break;
      }
      case TokenFit.START_NEW_REGION: {
        currentRegion = Region(
          start: token.start,
          end: token.end,
          tokens: [token],
        );
        regions.add(currentRegion);
        if (token.type == TOKEN_TYPE.DECIMAL) {
          currentRegion.hasDecimal = true;
        }
        break;
      }
      case TokenFit.NOPE:
      default: {
        currentRegion = null;
        break;
      }
    }
    i++;
  }

  regions.forEach((region) {
    region.subRegions = getSubRegions(region, options);
  });

  return regions;
}

TOKEN_TYPE getTokenType(String chunk) {
  if (UNIT_KEYS.contains(chunk.toLowerCase())) return TOKEN_TYPE.UNIT;
  if (TEN_KEYS.contains(chunk.toLowerCase())) return TOKEN_TYPE.TEN;
  if (MAGNITUDE_KEYS.contains(chunk.toLowerCase())) return TOKEN_TYPE.MAGNITUDE;
  if (DECIMALS.contains(chunk.toLowerCase())) return TOKEN_TYPE.DECIMAL;
  return null;
}

const puncExcludingDoubleQuote = r"!\#$%&'()*+,\-./:;<=>?@\[\\\]^_‘{|}~";

List<Region> parse(String text, Options options) {
  final chunks = splitKeepDelimeters(text, new RegExp('(\\w+|\\s|["$puncExcludingDoubleQuote])', caseSensitive: false));
  final tokens = List<Chunk>();
  for (String chunkText in chunks) {
    final unfuzzyChunk = chunkText.isNotEmpty && options.fuzzy && !PUNCTUATION.contains(chunkText) ?
        fuzzyMatch(chunkText) :
        chunkText;
    final start = tokens.isNotEmpty ? tokens[tokens.length - 1].end + 1 : 0;
    final end = start + chunkText.length;
    if (end != start) {
      tokens.add(Chunk(
        start: start,
        end: end - 1,
        value: unfuzzyChunk,
        lowerCaseValue: unfuzzyChunk.toLowerCase(),
        type: getTokenType(unfuzzyChunk),
      ));
    }
  }
  final regions = matchRegions(tokens, options);
  return regions;
}

List<String> splitKeepDelimeters(String text, RegExp regex) {
  final matches = regex.allMatches(text).toList();
  final result = List<String>();
  if (matches.isEmpty) {
    return [text];
  }
  if (matches[0].start > 0) {
    result.add(text.substring(0, matches[0].start));
  }
  int addedUpToIndex = matches[0].start;
  for (var match in matches) {
    if (addedUpToIndex < match.start) {
      result.add(text.substring(addedUpToIndex, match.start));
    }
    result.add(text.substring(match.start, match.end));
    addedUpToIndex = match.end;
  }
  if (matches.last.end != text.length) {
    result.add(text.substring(matches.last.end));
  }
  return result;
}

class Chunk {
  final int start;
  final int end;
  final String value;
  final String lowerCaseValue;
  TOKEN_TYPE type;

  Chunk({@required this.start, @required this.end, @required this.value, @required this.lowerCaseValue, @required this.type,});
}

class Region {
  int start;
  int end;
  List<Chunk> tokens;

  Region({@required this.start, @required this.end, @required this.tokens,});

  List<SubRegion> subRegions = [];
  bool hasDecimal = false;
}

class SubRegion {
  TOKEN_TYPE type;
  final List<Chunk> tokens;

  SubRegion({@required this.type, @required this.tokens});
}



lib/src/compiler.dart
import 'dart:math' as math;

import 'constants.dart';
import 'parser.dart';
import 'util.dart';

String getNumber(Region region) {
  int sum = 0;
  bool decimalReached = false;
  List<SubRegion> decimalUnits = [];
  region.subRegions.forEach((subRegion) {
    final type = subRegion.type;
    if (type == TOKEN_TYPE.DECIMAL) {
      decimalReached = true;
      return;
    }
    if (decimalReached) {
      decimalUnits.add(subRegion);
      return;
    }
    sum += subRegionSum(subRegion);
  });
  if (decimalUnits.isEmpty) {
    return sum.toString();
  } else {
    final allDecimalTokensAreUnits = decimalUnits.every((decimalSubRegion) {
      return decimalSubRegion.tokens.every((token) {
        return UNIT_KEYS.contains(token.lowerCaseValue);
      });
    });
    if (allDecimalTokensAreUnits) {
      double decimalSum = sum.toDouble();
      int currentDecimalPlace = 1;
      decimalUnits.forEach((decimalSubRegion) {
        decimalSubRegion.tokens.forEach((token) {
          decimalSum += NUMBER[token.lowerCaseValue] / math.pow(10, currentDecimalPlace);
          currentDecimalPlace += 1;
        });
      });
      return decimalSum.toString();
    } else {
      int decimalIntSum = 0;
      decimalUnits.forEach((decimalSubRegion) {
        decimalIntSum += subRegionSum(decimalSubRegion);
      });
      double decimalSum = (decimalIntSum / math.pow(10, decimalIntSum.toString().length));
      return (sum.toDouble() + decimalSum).toString();
    }
  }
}

int subRegionSum(SubRegion subRegion) {
  int subRegionSum = 0;
  final tokens = subRegion.tokens, type = subRegion.type;
  switch (type) {
    case TOKEN_TYPE.MAGNITUDE:
    case TOKEN_TYPE.HUNDRED: {
      subRegionSum = 1;
      final tokensCount = tokens.length;
      final acc = List<int>();
      tokens.asMap().forEach((i, token) {
        if (token.type == TOKEN_TYPE.HUNDRED) {
          final List<Chunk> tokensToAdd = (tokensCount - 1) > 0 ? tokens.sublist(i + 1) : [];
          final List<Chunk> filtered = [];
          for (int j = 0; j < tokensToAdd.length; j++) {
            if (j == 0 || tokensToAdd[j - 1].type.value > tokensToAdd[j].type.value) {
              filtered.add(tokensToAdd[j]);
            }
          }
          int tokensToAddSum = 0;
          filtered.forEach((tokenToAdd) => tokensToAddSum += NUMBER[tokenToAdd.lowerCaseValue]);
          return acc.add(tokensToAddSum + (NUMBER[token.lowerCaseValue] * 100));
        }
        if (i > 0 && tokens[i - 1].type == TOKEN_TYPE.HUNDRED) return;
        if (
          i > 1 &&
          tokens[i - 1].type == TOKEN_TYPE.TEN &&
          tokens[i - 2].type == TOKEN_TYPE.HUNDRED
        ) return;
        acc.add(NUMBER[token.lowerCaseValue]);
      });
      acc.forEach((tokenValue) {
        subRegionSum *= tokenValue;
      });
      break;
    }
    case TOKEN_TYPE.UNIT:
    case TOKEN_TYPE.TEN: {
      tokens.forEach((token) {
        subRegionSum += NUMBER[token.lowerCaseValue];
      });
      break;
    }
    default: 
    // no operation
  }
  return subRegionSum;
}

String replaceRegionsInText(List<Region> regions, String text) {
  String replaced = text;
  int offset = 0;
  regions.forEach((region) {
    final length = region.end - region.start + 1;
    final replaceWith = '${getNumber(region)}';
    replaced = splice(replaced, region.start + offset, length, replaceWith);
    offset -= length - replaceWith.length;
  });
  return replaced;
}

String compile(List<Region> regions, String text) {
  if (regions == null || regions.isEmpty) return text;
  if (regions[0].end - regions[0].start == text.length - 1) return getNumber(regions[0]).toString();
  return replaceRegionsInText(regions, text);
}


lib/src/constants.dart

const Map<String, int> UNIT = {
  'zero': 0,
  'first': 1,
  'one': 1,
  'second': 2,
  'two': 2,
  'third': 3,
  'thirteenth': 13,
  'thirteen': 13,
  'three': 3,
  'fourth': 4,
  'fourteenth': 14,
  'fourteen': 14,
  'four': 4,
  'fifteenth': 15,
  'fifteen': 15,
  'fifth': 5,
  'five': 5,
  'sixth': 6,
  'sixteenth': 16,
  'sixteen': 16,
  'six': 6,
  'seventeenth': 17,
  'seventeen': 17,
  'seventh': 7,
  'seven': 7,
  'eighteenth': 18,
  'eighteen': 18,
  'eighth': 8,
  'eight': 8,
  'nineteenth': 19,
  'nineteen': 19,
  'ninth': 9,
  'nine': 9,
  'tenth': 10,
  'ten': 10,
  'eleventh': 11,
  'eleven': 11,
  'twelfth': 12,
  'twelve': 12,
  'a': 1,
};

const Map<String, int> TEN = {
  'twenty': 20,
  'twentieth': 20,
  'thirty': 30,
  'thirtieth': 30,
  'forty': 40,
  'fortieth': 40,
  'fifty': 50,
  'fiftieth': 50,
  'sixty': 60,
  'sixtieth': 60,
  'seventy': 70,
  'seventieth': 70,
  'eighty': 80,
  'eightieth': 80,
  'ninety': 90,
  'ninetieth': 90,
};

const Map<String, int> MAGNITUDE = {
  'hundred': 100,
  'hundredth': 100,
  'thousand': 1000,
  'million': 1000000,
  'billion': 1000000000,
  'trillion': 1000000000000,
  'quadrillion': 1000000000000000,
  'quintillion': 1000000000000000000,
};

const Map<String, int> NUMBER = { ...UNIT, ...TEN, ...MAGNITUDE };

final List<String> UNIT_KEYS = UNIT.keys.toList();
final List<String> TEN_KEYS = TEN.keys.toList();
final List<String> MAGNITUDE_KEYS = MAGNITUDE.keys.toList();

final List<String> NUMBER_WORDS = [ ...UNIT_KEYS, ...TEN_KEYS, ...MAGNITUDE_KEYS];

const List<String> JOINERS = ['and'];
const List<String> DECIMALS = ['point', 'dot'];

const List<String> PUNCTUATION = [
  '.',
  ',',
  '\\',
  '#',
  '!',
  '\$',
  '%',
  '^',
  '&',
  '/',
  '*',
  ';',
  ':',
  '{',
  '}',
  '=',
  '-',
  '_',
  '`',
  '~',
  '(',
  ')',
  ' ',
];

enum TOKEN_TYPE {
  UNIT,
  TEN,
  MAGNITUDE,
  DECIMAL,
  HUNDRED,
}

extension TypeExtension on TOKEN_TYPE {

  int get value {
    switch (this) {
      case TOKEN_TYPE.UNIT:
        return 0;
      case TOKEN_TYPE.TEN:
        return 1;
      case TOKEN_TYPE.MAGNITUDE:
        return 2;
      case TOKEN_TYPE.DECIMAL:
        return 3;
      case TOKEN_TYPE.HUNDRED:
        return 4;
    }
    throw "Null TOKEN_TYPE";
  }

  void talk() {
    print('meow');
  }
}

final List<String> ALL_WORDS = [ ...NUMBER_WORDS, ...JOINERS, ...DECIMALS ];

const List<String> BLACKLIST_SINGULAR_WORDS = ['a'];
